primitives/src/predicate.rs

```diff
@@ -0,0 +1,333 @@
+use crate::{Claim, ClaimType, Rule, RuleType};
+use codec::{Decode, Encode};
+
+use sp_std::prelude::*;
+
+/// Context using during an `Predicate` evaluation.
+///
+/// # TODO
+///  - Use a lazy access to claims. It could be part of the optimization
+///  process of CDD claims.
+#[derive(Encode, Decode, Clone, Debug, Default)]
+pub struct Context {
+    /// Predicate evaluation will use those claims.
+    pub claims: Vec<Claim>,
+}
+
+impl From<Vec<Claim>> for Context {
+    fn from(claims: Vec<Claim>) -> Self {
+        Context { claims }
+    }
+}
+
+// Predicate Trait
+// ==================================
+
+/// It allows composition and evaluation of claims based on a context.
+pub trait Predicate {
+    /// It evaluates this predicated based on `context` context.
+    fn evaluate(&self, context: &Context) -> bool;
+
+    /// It generates a new predicate that represents the logical AND
+    /// of two predicates: `Self` and `other`.
+    #[inline]
+    fn and<B>(self, other: B) -> AndPredicate<Self, B>
+    where
+        Self: Sized,
+        B: Predicate + Sized,
+    {
+        AndPredicate::new(self, other)
+    }
+
+    /// It generates a new predicate that represents the logical OR
+    /// of two predicates: `Self` and `other`.
+    #[inline]
+    fn or<B>(self, other: B) -> OrPredicate<Self, B>
+    where
+        Self: Sized,
+        B: Predicate + Sized,
+    {
+        OrPredicate::new(self, other)
+    }
+
+    /// It generates a new predicate that represents the logical NOT
+    /// of this predicate.
+    #[inline]
+    fn not(self) -> NotPredicate<Self>
+    where
+        Self: Sized,
+    {
+        NotPredicate::new(self)
+    }
+}
+
+// Helper functions
+// ======================================
+
+/// It creates a predicate to evaluate the existential of `claim` in the context.
+#[inline]
+pub fn exists(claim_type: ClaimType) -> ExistentialPredicate {
+    ExistentialPredicate { claim_type }
+}
+
+/// It creates a predicate to evaluate if any of `claims` are found in the context.
+#[inline]
+pub fn any(claims: Vec<Claim>) -> AnyPredicate {
```

Comment:
I used any because it is quite similar to the concept of `iter::any`. 
I wouldn't use `Intersects` because it could implies to return the intersection instead of a boolean.